mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/util/HierarchyScanner.java

```diff
@@ -0,0 +1,241 @@
+      return;
+    }
+
+    Set<TypeMirror> interfacesToScan = Collections.emptySet();
+    boolean atHighestClass = hierarchyScanStrategy.atHighest(classElement.asType());
+    while (!atHighestClass) {
+      // add super class
+      TypeMirror superClass = classElement.getSuperclass();
+      TypeElement superClassElement = null;
+      if (superClass.getKind() == TypeKind.DECLARED) {
+        superClassElement = (TypeElement) context.getTypeUtils().asElement(superClass);
+        atHighestClass = hierarchyScanStrategy.atHighest(superClass);
+        if (!atHighestClass || hierarchyScanStrategy.includeHighestAncestor) {
+          if (!typeConsumer.apply(superClass)) {
+            return;
+          }
+        }
+      } else {
+        // at highest level, no need to proceed.
+        atHighestClass = true;
+      }
+
+      // as we encounter interfaces, also keep track of their parents.
+      Set<TypeMirror> newInterfacesToScan = Sets.newLinkedHashSet();
+
+      // scan parent classes interfaces and add them.
+      scanInterfaces(
+          hierarchyScanStrategy,
+          classElement.getInterfaces(),
+          newInterfacesToScan,
+          context,
+          typeConsumer);
+      // and then add interfaces to scan from previous class interfaces parents.
+      scanInterfaces(
+          hierarchyScanStrategy, interfacesToScan, newInterfacesToScan, context, typeConsumer);
+
+      // navigate up to the superclass and to the class' encountered interfaces' parents.
+      classElement = superClassElement;
+      interfacesToScan = newInterfacesToScan;
+    }
+
+    // if we've exhausted the class hierarchy, we may still need to consume the interface hierarchy.
+    while (!interfacesToScan.isEmpty()) {
+      Set<TypeMirror> newInterfacesToScan = Sets.newLinkedHashSet();
+      scanInterfaces(
+          hierarchyScanStrategy, interfacesToScan, newInterfacesToScan, context, typeConsumer);
+      interfacesToScan = newInterfacesToScan;
+    }
+  }
+
+  private static void scanInterfaces(
+      HierarchyScanStrategyOptions hierarchyScanStrategy,
+      Collection<? extends TypeMirror> interfacesToScan,
+      Set<TypeMirror> newInterfacesToScan,
+      ProcessorContext context,
+      Function<TypeMirror, Boolean> typeConsumer) {
+    for (TypeMirror interfaceType : interfacesToScan) {
+      if (interfaceType.getKind() == TypeKind.DECLARED) {
+        TypeElement interfaceElement =
+            (TypeElement) context.getTypeUtils().asElement(interfaceType);
+        // skip if at highest ancestor.
+        boolean atHighest = hierarchyScanStrategy.atHighest(interfaceType);
+        if (!atHighest || hierarchyScanStrategy.includeHighestAncestor) {
+          if (!typeConsumer.apply(interfaceType)) {
+            return;
+          }
+        }
+        if (!atHighest) {
+          newInterfacesToScan.addAll(interfaceElement.getInterfaces());
+        }
```

Comment:
1. If `atHighest` happens for an interface, we don't communicate that to the caller. Wouldn't it then continue to navigate superclasses?
2. If it happens for anything but the first interface, we've already added the parents of the previous interfaces at the same level. Example: assuming highest=I2, we will traverse I3 (and all its parents I think).
    ```
    I3
    |
    I1 I2
    |  /
     C
```

Reply:
> If atHighest happens for an interface, we don't communicate that to the caller. Wouldn't it then continue to navigate superclasses?

That's right, I had considered changing `highestAncestor` to `highestAncestors`, but after thinking about it some more realized that'd maybe be overthinking it.  Not sure we should expect very complex type hierarchies where you need to prevent scanning at both the class and interface level.  I could add that though if you think it is worth it.

> If it happens for anything but the first interface, we've already added the parents of the previous interfaces at the same level

Correct, that's the way I intended it.  If `I2` is highest, that effectively means don't scan anything above `I2`, since `I1` is a separate hierarchy, `I3` would also be scanned.

Reply:
> Not sure we should expect very complex type hierarchies

Agreed, in practice it will probably be a couple of levels of either parent classes or interfaces.

If the current behavior is as intended, that's fine with me as long as we document it somewhere in the manual.

Reply:
I added details about this to `highestAncestor` javadoc in `HierarchyScanStrategy`.  I don't cover `HierarchyScanStrategy` in the manual, other than to mention and link to the javadoc for it in both inheritance sections as I expect a large percentage of the time the default behavior should be sufficient, and the javadoc does a good job explaining how to use it, but let me know if you'd like me to add more to the manual.